aten/src/ATen/FunctionalStorageImpl.h

```diff
@@ -103,6 +103,31 @@ struct TORCH_API FunctionalStorageImpl : public c10::StorageImpl {
 
   ~FunctionalStorageImpl() override = default;
 
+  void mark_mutation() {
+    mutation_counter_++;
+  }
+  void mark_mutation_during_no_grad_or_inference_mode() {
+    mutation_counter_during_no_grad_or_inference_mode_++;
+  }
+  void mark_mutation_hidden_from_autograd() {
+    mutation_counter_hidden_from_autograd_++;
+  }
+
+  bool are_all_mutations_under_no_grad_or_inference_mode() const {
+    auto non_autograd_mutations =
+        mutation_counter_during_no_grad_or_inference_mode_ +
+        mutation_counter_hidden_from_autograd_;
+    // The <= is because both counters will technically be incremented, if we
+    // perform e.g. a triton kernel mutation under no_grad
+    return mutation_counter_ <= non_autograd_mutations;
+  }
+
+  bool are_all_mutations_hidden_from_autograd() const {
+    // mutations under no_grad / inference_mode are technically not hidden from
+    // autograd - they change the version counter
+    return mutation_counter_ <= mutation_counter_hidden_from_autograd_;
```

Comment:
is mutation_counter incremented each time we increment mutation_counter_hidden_from_autograd_?

Reply:
The only place this is used to today is for triton kernels (which hide all mutations from autograd).

There's a HOP for triton kernels, that calls this API [here](https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/triton_kernel_wrap.py#L808). And it's carefully written to bump both of those mutation counters by the same amount (which is an invariant we need to make sure we maintain in any future places where we need to call this API).

Reply:
Can we make illegal states unrepresentable? Instead of `mark_mutation` have `mark_mutation_sees_autograd`. Now they're all disjoint and you test if certain fields are zero or not.

Reply:
I guess part of the reason is that these were preexisting lol

Reply:
yeah, part of the problem is that the triton HOP (linked above) has to call `replace_()` in its functionalization rule, which will implicitly bump the `mutation_counter_`.

One option is that I add a kwarg to `replace_(hidden_from_autograd=False)`, so `replace_()` is the one API that everyone goes through to bump mutation counters.